### PR TITLE
#1045 - Batch Sort

### PR DESCRIPTION
**Pull request by hunyoboy**
Updates:

1. From submitting reorder updates every sort change, instead we batch the submission to max count.
2. For more responsive feel, edit of gallery sequence is being kept on the state and treated as a regular sort order event.
3. Added a “reorder images" button with a counter for total changes made in the gallery sequence.
4. For better user experience, adding/deleting images remembers the previous unsaved sort changes.
5. Added an auto-correction algorithm where duplication of gallery sequence issue is prevented.

### CODE REVIEW

**Comment by hunyoboy** [comment]
@joshlevy Can you take a look if you have a moment? Thank you.

File: src/plugins/dam/sagas/reorderGalleryAssetsFlow.js [hunk 1]
```diff
@@ -87,16 +87,15 @@ export default function* reorderGalleryAssetsFlow({
...
-  yield fork([toast, 'show']);
-  yield putResolve(pbj);
-  const result = yield race({
-    event: call(waitForMyEvent, eventChannel, config.assetCount),
-    timeout: delay(7500),
-  });
-
   try {
+    const expectedEvent = config.schemas.galleryAssetReordered.getCurie().toString();
+    const eventChannel = yield actionChannel(expectedEvent, buffers.dropping(10));
```
Comment (gdbrown) [inline src/plugins/dam/sagas/reorderGalleryAssetsFlow.js]: @hunyoboy let's not drop any events, remove the buffering

File: src/plugins/dam/sagas/reorderGalleryAssetsFlow.js [hunk 2]
```diff
@@ -87,16 +87,15 @@ export default function* reorderGalleryAssetsFlow({
...
+    yield fork([toast, 'show']);
+    yield putResolve(pbj);
+    const result = yield race({
+      event: call(waitForMyEvent, eventChannel, config.assetCount),
+      timeout: delay(7500),
```
Comment (gdbrown) [inline src/plugins/dam/sagas/reorderGalleryAssetsFlow.js]: @hunyoboy timeout here should be base of 5000 but expand as assetCount increases

File: src/plugins/curator/components/gallery-media/utils/moveNodeByGallerySequence.test.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
+import test from 'tape';
+import isEqual from 'lodash/isEqual';
+import AssetV1Mixin from '@triniti/schemas/triniti/dam/mixin/asset/AssetV1Mixin';
+import AssetId from '@triniti/schemas/triniti/dam/AssetId';
+
+import moveNodeByGallerySequence from './moveNodeByGallerySequence';
+
+const assetSchemas = AssetV1Mixin.findAll();
+const imageAssetSchema = assetSchemas[4];
```
Comment (gdbrown) [inline src/plugins/curator/components/gallery-media/utils/moveNodeByGallerySequence.test.js]: @hunyoboy tests should just pull in the acme schema you want direct.  references array index from findAll is fragile.  if that's not doable then use the image asset mixin to do the resolution, then you can use findOne

File: src/plugins/curator/components/gallery-media/index.jsx [hunk 4]
```diff
@@ -79,58 +79,52 @@ class GalleryMedia extends React.Component {
...
     const { delegate } = this.props;
-    await delegate.handleAddGalleryAssets(assetMap);
+    try {
+      await delegate.handleAddGalleryAssets(assetMap);
+    } catch (e) {
+      // did not update, should restore tab
+    }
     delegate.handleSearchGalleryAssets();
 
     return Promise.resolve();
```
Comment (gdbrown) [inline src/plugins/curator/components/gallery-media/index.jsx]: @hunyoboy an async function already returns a promise.  you don't need to make one yourself

**Pull request merged**